### ISSUE TITLE
deprecated lint should not warn within a deprecated module

### ISSUE THREAD

**Issue by aturon** [rust-lang/rust#16490]
The stability lint was recently changed to warn on use of deprecated items even when they are defined in the same crate. An unintended consequence is that, when an entire module is deprecated, the body of the module will warn on every use of items defined within it.

We should change this behavior, but it's not entirely clear what the rule should be or how to implement it.

**Comment by steveklabnik** [comment]
To reproduce:

```
#![feature(staged_api)]
#![staged_api]

#[unstable(feature = "lol", issue = "1")]
#[deprecated(since = "1.0.0", reason = "lol")]
mod nope {
    pub fn foo() {}
    pub fn bar() {}
}

fn main() {
    nope::foo();
    nope::bar();
}
```

this will give both

```
hello.rs:12:5: 12:14 warning: use of deprecated item: lol, #[warn(deprecated)] on by default
hello.rs:12     nope::foo();
                ^~~~~~~~~
hello.rs:13:5: 13:14 warning: use of deprecated item: lol, #[warn(deprecated)] on by default
hello.rs:13     nope::bar();
                ^~~~~~~~~
```

It's been a year, with no discussion at all. I'm not sure this is worth doing.

**Comment by TimNN** [comment]
This just came up again in #35128, although a little bit more surprisingly, due to the expansion of `format_args!` inside of a deprecated function.

@steveklabnik: If I understand the original issue correctly, I believe your code should continue to warn about the `use of deprecated item`.

I believe the issue is that the following warning should not happen:

```rust
#[deprecated]
mod nope {
    fn foo() {}

    fn bar() {
        foo() //~ WARN use of deprecated item
    }
}

**Comment by eddyb** [comment]
The fix doesn't seem *that* hard. [`depr_map`](https://github.com/rust-lang/rust/blob/5556554e2504ba2df8712ad0666ce1a11915dade/src/librustc/middle/stability.rs#L69) can contain a `DefId` of the definition marked with `#[deprecated]` instead of `Deprecation` (which [can be obtained on demand](https://github.com/rust-lang/rust/blob/5556554e2504ba2df8712ad0666ce1a11915dade/src/librustc/middle/stability.rs#L180)).
Then, if the current node and the referenced node both have the same `DefId`, no warning is issued.

**Comment by TimNN** [comment]
If nobody has picked this up by then, I'll see if I can come up with a fix for this towards the end of the week.

**Comment by cengizIO** [comment]
@eddyb I can work this if nobody else wants to.

LiveEdit: I guess @TimNN wants. 🍀
